fat/fat2/transaction.go

```diff
@@ -0,0 +1,168 @@
+package fat2
+
+import (
+	"fmt"
+)
+
+// Transaction represents a fat2 transaction, which can be a normal account
+// transaction or a conversion transaction depending on the Inputs and Outputs
+type Transaction struct {
+	Input    TransactionIO   `json:"input"`
+	Outputs  []TransactionIO `json:"outputs"`
+	Metadata interface{}     `json:"metadata,omitempty"`
+}
+
+// Valid performs all validation checks and returns nil if t is a valid Transaction
+func (t *Transaction) Valid() error {
+	// Perform initial validation on all inputs and outputs in isolation
+	if err := t.Input.Valid(); err != nil {
+		return fmt.Errorf("invalid input: %v", err)
+	}
+	if t.Input.Address == nil {
+		return fmt.Errorf("invalid input: no address")
+	} else if t.Input.Address.String() == "FA1zT4aFpEvcnPqPCigB3fvGu4Q4mTXY22iiuV69DqE1pNhdF2MC" {
+		return fmt.Errorf("invalid input: FA1zT4aFpEvcnPqPCigB3fvGu4Q4mTXY22iiuV69DqE1pNhdF2MC reserved as burn address")
+	}
+	if t.Input.Type == nil {
+		return fmt.Errorf("invalid input: no token type")
+	}
+	if len(t.Outputs) == 0 {
+		return fmt.Errorf("at least one output required")
+	}
+	for i, output := range t.Outputs {
+		if err := output.Valid(); err != nil {
+			return fmt.Errorf("invalid output at index %d: %v", i, err)
+		}
+	}
+
+	// TODO: Perform contextual validation on the collection of inputs and outputs
+	inputHasAmount := t.Input.Amount != nil
+	for i, output := range t.Outputs {
+		outputHasAddress := output.Address != nil
```

Comment:
See my comment in transactionio.go about doing logic this way.

Reply:
This section makes me conscious of the fact that this transaction type is very flexible. it can represent many kinds of transactions. That makes me concerned as it means the code is less precise. Idk what can be done about that...

Reply:
Given the complexity of this logic I'd like to hear in words what has to be validated. I wonder if there is not a simpler way to do this logic. I can't say there is but I wonder if we can think about how to make this more clear and add confidence that there are not mistakes when initially reading it.

Reply:
Refactored and eliminated this logic